### ISSUE TITLE
As a dev, I need to review the usefulness and presence of some functions in Auth-browser

### ISSUE THREAD

**Issue by tmodoux** [pryv/lib-javascript#40]
Some functions make maybe more sense in pryv.Connection:
logout, login, trustedLogout, whoAmI, loginWithCookie

**Comment by sgoumaz** [comment]
I think it's clearer from both outside and inside to keep functions related to user authentication in `Auth`. (I even kind of remember some of those used to be in `Connection` but we moved them away.) I totally understand why it seems to make more sense to put them together, but our API is different from most (as things currently stand) in that a connection can be established not only by authenticating as a user (auth process, username/password), but also by simply providing an access token (a 'bearer token' in general security terms).

A couple of thoughts that may help (feel free to amend the docs if you feel it's lacking) (and to tell me if I don't make sense):

- `Auth` handles user authentication, i.e. authenticating with a "personal" access. When the auth process succeeds a session is opened for the personal access (which is thus given a non-permanent access token). _Then_ a connection can be established.
- So `Connection` handles what happens from the moment you do have an access token. You can think of a connection as an 'instanciated access' from the client's perspective. It's the handle to actually interacting with user data.

**Comment by tmodoux** [comment]
Thanks for your clarification, this makes now sense for me for login, logout, whoAmI and loginWithCookie.

But I am still not 100% confident about trustedLogout.
Could you please give me an use case for this function? So for you it makes no sense in Connection, right?

**Comment by sgoumaz** [comment]
Hmmm… to be frank I did not know/remember about `trustedLogout()`. Looking at the code (also `Connection`'s) makes me wonder as well… a few things don't feel right to me. Suggestions in order to keep things as I described above:

- Move everything related to `/auth` API calls to `Auth`; that includes `Connection.login()` and `trustedLogout()` (please do fix the very anti-node callback scheme on that one) which will become just `logout()` to keep in line with the API (the existing `logout()` should be renamed as it's confusing).
- So for example `Auth.login()` will do the API call to `/auth/login` then instantiate a `Connection` if successful
- And instead of `Connection.trustedLogout()` make an access type-agnostic `Connection.close()` which will delegate to `Auth.logout()` if the access is "personal" (I'm not 100% sure about this one, may have to be discussed)

**Comment by tmodoux** [comment]
I just brought back trustedLogout to Auth and renamed both confusing functions as you suggested.
You can track these changes here: [https://github.com/pryv/lib-javascript/tree/refactor/auth-functions](https://github.com/pryv/lib-javascript/tree/refactor/auth-functions)
I think that keeping Connection.trustedLogout (or an equivalent access type-agnostic Connection.close()) is not necessary since trustedLogout was initially in Auth and thus never used in Connection (my mistake to put it in Connection).

About the node callback scheme, do you think about something like that:
```
Auth.prototype.logout = function(callback) {
  if (this.connection) {
    this.connection.request({
      method: 'POST',
      path: '/auth/logout',
      callback: callback
    });
  }
};
```
If so, it would require to refactor the way to provide callbacks to these functions, and get rid of the callback's array: settings.callbacks ? Besides, almost all Auth functions are using this callback scheme, which seem strange to me...

Now about login, I am less confident to move it completely to Auth (and remove it from Connection). Is there not some external services using Connection.login without using Auth? How can we handle this transition?

**Comment by sgoumaz** [comment]
Thanks Thiébaud.

About the node-style callbacks, yes I mean async functions taking a single callback, and calling it with an error (or null) as the first argument. If that brings down the whole tapestry and is too much for this issue's scope, please write it down in another one for follow up.

About `login()` I'm not sure I get it… any client code currently calling `Connection.login()` can surely just replace that with `Auth.login()`? Note that we certainly need a transitional phase in which `Connection.login()` still exists to output a "deprecated" warning and delegate to `Auth.login()`.

**Comment by tmodoux** [comment]
Yes then I think refactoring callbacks in a node-style manner will bring enough changes to be a separate issue.

You got it right for the login(), I was thinking about this kind of transitional phase :)

Thanks for your help.

**Comment by kschiess** [comment]
@tmodoux Can you close this if it is answered with enough clarity? Or - in the opposite case - make it clear for everyone what needs be done, in the form of a small summary perhaps?
